### ISSUE TITLE
UI for desktop port

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#21333]
We eventually want to have a minimal UI for the desktop port.

This would include a toolbar and a list of tabs.

The constrains we have are:
- multiplatform (Linux/Windows/Mac)
- in Rust or JS (we want to limit the amount of toolchain to support)
- no intrusive changes in the engine (as in, under /components/) or web standards bending

Non goals:
- Innovative UI: we are not trying to build a next generation user interface. The UI will just be here to make it easier to test servo.
- Mobile support: we already have an android port.

MVP:
- tabs
- multi windows support
- back, forward, reload, stop buttons and urlbar

## Suggested approach

This requires 2 big changes:

### 1. Graphic:

- draw 2 WR documents in one glutin window: chrome and content
- content WR document will display multiple top level pipelines, aka tabs (already supported)
- chrome WR document will load one single HTML document, the UI

### 2. Controls:

ports/servo/ will be in charge of all the wiring. All events from the content (see `EmbedderMsg`) will be used to update the state of the browser, in Rust. This state will forwarded to the chrome document as it changes. The chrome document will be able to send message back to ports/servo/.

Example: **State forwarding: port to chrome document**. If the title of a web page changes, a Rust `EmbedderMsg` is sent to ports/servo, which will update the state of the browser. The state will then be forwarded to the chrome document, which will update its UI to reflect the new title. **Message passing: chrome document to port**. If the user clicks on the reload button, the chrome document will send a message to ports/servo that will perform the reload.

The state forwarding and the message passing mechanisms between the chrome document and ports/main still need to be defined. We don't want to introduce security issues and we don't want to invent a new standard. A possible approach would be to rely on postMessage. 

## Difference with browserhtml

Even though the buttons and url will be drawn in HTML, this approach is different from browserhtml. We are not introducing new JS APIs, we do not introduce new HTML elements, we do not use iframes or webviews or mozbrowsers, and the browser logic is still written in Rust. The HTML/JS part will only be here to draw the UI based on the state sent from ports/servo.

## Status

Nothing has been done yet for the controls part.

For the graphic part, WR already supports multiple documents (I believe this is how it works in Firefox+WR), but Servo's compositor doesn't. I have a WIP branch that introduces areas, which are WR documents in one GL buffer. See previous discussion about tab support: https://github.com/servo/servo/issues/20335

For multi windows support, still more work needs to be done at the compositor level: https://github.com/servo/servo/issues/13997

Screenshot of a WIP prototype:

![zind1jm](https://user-images.githubusercontent.com/373579/43626322-aaadf59e-96f1-11e8-9d08-ea608c43add2.png)

**Comment by fabricedesre** [comment]
Nit picking on "the browser logic is still written in Rust" : there will be some logic in JS loaded in the chrome document. I predict that it will grow more than you'd like :slightly_smiling_face:

**Comment by paulrouget** [comment]
Right. But hopefully not too much because the JS code has access to nothing but the webpage and the browser state. Also we are not trying to build a full-featured browser. But yeah, I've seen how these things get bloated :)

**Comment by tigercosmos** [comment]
What do the Servo team think about this?

**Comment by jdm** [comment]
I'm in favour of the design proposed by Paul.

**Comment by fabricedesre** [comment]
Let me recap a bit:
- the chrome WR doc will render eg. the address bar and tap strip using HTML/CSS/js.
- the Rust embedder will position content WR Docs, making all layout calculation to be positioned and sized properly and (usually) not overlap with the chrome doc.
- when something like window.alert("Servo!") is called, the Rust embedder decides how to handle it. This usually is tab-modal, so needs to be drawn on top of the content doc caller. It seems to me that this rules out rendering that by the chrome WR doc, and mandates rendering using a "native" toolkit instead.

That looks complicated, and I'm worried that it will be hard to achieve eg. consistent animations when using different rendering stacks.  I struggle to see how useful the chrome document is since it can't be used to render all the chrome UI.

**Comment by paulrouget** [comment]
@fabricedesre All your concerns are valid if you're thinking about building a full browser. A full browser needs sidebar, toolbars, statusbar, popups, preferences pane, animations, integration with the OS, the awesomebar, etc… And that would be a lot of messy work with the approach I'm proposing. A better approach would be to do what I did with servoshell. Native UI + servo as a lib.

But like I mentioned in the first comment, **the UI will just be here to make it easier to test servo**.

So we're going for the minimal approach that requires the least amount of extra work, doesn't require a 3rd party toolkit, an extra build system or a new language.

To address your specific concerns:

> the Rust embedder will position content WR Docs, making all layout calculation

The calculation will be very minimal as it's just one bar on top of the content.

> when something like window.alert

We can have a 3rd document or even just keep tinyfiledialogs.

---

Another thing I care about is to keep the logic in Rust. Maybe as a future crate people could reuse to build their own browser. Hopefully, if we can manage to build the exact mechanism I suggested (state forwarding and message passing) anybody could be their own UI following this protocol. I used this approach for the HTML UI, the ServoShell Cocoa UI in C#, and the miniui. All different toolkits and even languages.

**Comment by paulrouget** [comment]
Having a transparent top document is totally possible. It's maybe even better if we want to show popups & co.

> I guess I would still need to figure out how to let events be targeted properly

There's nothing specially tricky here. There is no notion of focus. It's up to the embedder to decide if the content is focused or not, and then send the events to the right browser. As for mouse events, a hittest will do.

**Comment by tigercosmos** [comment]
@paulrouget  Is the screenshot of the WIP prototype above on the way? or just drawn by PS?
Also, what is "chrome WR document", and what does the word "chrome" means here?

**Comment by paulrouget** [comment]
@tigercosmos it’s a screenshot 

WR document is an area on the screen where servo paint a set of pages.

Chrome refers to a document that will be used to draw the UI.

**Comment by fabricedesre** [comment]
@paulrouget do you have any schedule / ETA ? Is that still the plan to do that?

**Comment by paulrouget** [comment]
It is still planned. No ETA yet. Do you need multi-document support soon?

**Comment by fabricedesre** [comment]
This would be a lie to claim it's urgent :) I'm just tired of rebasing my hacked <iframe webview> patches so I would not mind testing this setup.

**Comment by jdm** [comment]
We have a desktop UI now, so I'm going to close this.
